### PR TITLE
ICS28: Consumer Initiated Slashing (aka Evidence)

### PR DESCRIPTION
**Pull request by mpoke**
Closes https://github.com/cosmos/ibc/issues/685

The partial specification of Cross-Chain Validation.

Changes to base (i.e., [ccv](https://github.com/cosmos/ibc/tree/marius/ccv-init-genesis)):
- Consumer Initiated Slashing, i.e., the evidence / slashing subprotocol that enables the provider chain to punish validators for misbehaviors on the consumer chain.
- Mapping of consumer heights to provider heights, which is needed by the Consumer Initiated Slashing to set the evidence `infractionHeight`.
- `VSCMaturedPacket` instead of `VSCpacket` acknowledgements to notify the provider chain when unbonding operations can complete (i.e., reached maturity on the consumer chain). This is needed to keep the order between this notifications and the slashing requests. 

Features not included (to be added to other PRs):
- The distribution subprotocol that enables the validators to receive rewards for validating on the consumer chain.

### CODE REVIEW

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 1]
```diff
@@ -205,3 +213,53 @@ This approach is depicted in the following Figure that shows an overview of the
...
+then the tokens MUST NOT be slashed.
+
+In the context of CCV, validators (with tokens bonded on the provider chain) MUST be slashed for infractions committed on the consumer chains. 
+Thus, although the infractions are committed on the consumer chains and evidence of these infractions is submitted to the consumer chains, the slashing happens on the provider chain. As a result, the Consumer Initiated Slashing operation requires, for every consumer chain, a mapping from consumer chain block heights to provider chain block heights.
+The following figure shows an overview of how CCV creates such a mapping using VSC IDs.
+For clarity, we use `Hp*` and `Hc*` to denote block heights on the provider chain and consumer chain, respectively. 
+
+![Mapping Between Provider and Consumer Heights](./figures/ccv-height-mapping-overview.png?raw=true)
+
+- For every block, the provider CCV module maps the ID of the VSC it provides to the consumer chains to the height of the subsequent block, i.e., `VSCtoH(VSC.id) = Hp + 1`, for a VSC provided at height `Hp`.
```
Comment (jtremback) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:241]: I had a hard time understanding the notation here. To me it sounds like there is a map data structure of some sort looking like this:

```
{
  [VSC.id]: Hp + 1,
  ...
}
```

`VSCtoH(VSC.id) = Hp + 1` implies to me some kind of conversion or function call. Can we use square brackets like this? `VSCtoH[VSC.id] = Hp + 1` That is more in line with the map accessing syntax in most languages.

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 2]
```diff
@@ -205,3 +213,53 @@ This approach is depicted in the following Figure that shows an overview of the
...
+
+- At height `Hc2`, the consumer chain receives evidence that a validator `V` misbehaved at height `Hc1`. 
+  As a result, the consumer CCV module sends a `SlashPacket` to the provider chain: 
+  It makes a request to slash `V`, but it replaces the infraction height `Hc1` with `HtoVSC(Hc1)`, 
+  i.e., the ID of the VSC that updated the "misbehaving voting power" or `0` if such a VSC does not exist.
+- The provider CCV module receives at height `Hp1` the `SlashPacket` with `vscId = HtoVSC(Hc1)`. 
+  As a result, it requests the provider Slashing module to slash `V`, but it set the infraction height to `VSCtoH(vscId)`, i.e., 
+    - if `vscId != 0`, the height on the provider chain where the voting power was updated by the VSC with ID `vscId`;
+    - otherwise, the height at which the first VSC was provided to this consumer chain.
+  > **Note**: As a consequence of slashing (and potentially jailing) `V`, the Staking module updates accordingly `V`'s voting power. This update MUST be visible in the next VSC provided to the consumer chains.
```
Comment (jtremback) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:263]: Is this MUST requirement enforced anywhere or is it simply a clarification for the reader?

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 3]
```diff
@@ -205,3 +213,53 @@ This approach is depicted in the following Figure that shows an overview of the
...
+
+> **Note**: In the context of single-chain validation, usually `evidenceHeight = slashingHeight`. 
+
+The [Security Model](#security-model) guarantees that any misbehaving validator can be slashed for at least the unbonding period, 
+i.e., as long as that validator's tokens are not unbonded yet, they can be slashed. 
+However, if the tokens start unbonding before `infractionHeight`, i.e., the tokens did not contribute to the voting power that committed the infraction, 
+then the tokens MUST NOT be slashed.
+
+In the context of CCV, validators (with tokens bonded on the provider chain) MUST be slashed for infractions committed on the consumer chains. 
+Thus, although the infractions are committed on the consumer chains and evidence of these infractions is submitted to the consumer chains, the slashing happens on the provider chain. As a result, the Consumer Initiated Slashing operation requires, for every consumer chain, a mapping from consumer chain block heights to provider chain block heights.
```
Comment (jtremback) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:235]: Could use a diagram of the mapping as a supplement to the current diagram that illustrates more of the mechanic. Something really simple like 
```
        Consumer                  IBC               Provider

cHeight <--- VSCtoCHeight[] ---> VSCid <--- VSCtoPHeight[] ---> pHeight
```

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md (hunk 2 above)
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:263]: Here is just a not for the reader, but it's actually an assumption of the ABCI application (i.e., the state machine). The test case should be in the Staking module of Cosmos SDK (i.e., if during a block the voting power changes due to a Slash operation, at the end of the block, that change must be mirrored in the list of updates provided to the underlying consensus engine).

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md (hunk 3 above)
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:235]: I'm not sure I understand the diagram. From the names, both mappings seems to map from VSCs to heights. From the arrows, I'm not sure what's the input and what's the output. Also, why is vscId under IBC?  

Do you mean sth like this?
```
        Consumer                                Provider

cHeight ---> HtoVSC[] ---> VSCid ---> VSCtoH[] ---> pHeight
```
I don't really see how this is more helpful than the "Mapping consumer heights to provider heights" figure and its explanation.

Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:235]: @jtremback I added an extra diagram that provides the intuition  behind the mapping.

**Comment by mpoke** [comment]
Commit https://github.com/cosmos/ibc/pull/663/commits/e0401ee4c307318af88e39903a25e24109b3dcef deals with issue https://github.com/cosmos/ibc/issues/685

**Review by AdityaSripal — COMMENTED** [review]
Part way through

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 4]
```diff
@@ -205,3 +213,62 @@ This approach is depicted in the following Figure that shows an overview of the
...
+
+The following figure shows the intuition behind such a mapping using the provided VSCs. 
+The four unbonding operations (i.e., undelegations) occur on the provider chain and, as a consequence, the provider chain provides VSCs to the consumer chain, e.g., `undelegate-3` results in `VSC3` being provided.
+The four color (i.e., red, blue, green, and yellow) indicate the mapping of consumer chain heights to provider chain heights (note that on the provider chain there is only one block of a given color).
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:239]:
```suggestion
The four colors (i.e., red, blue, green, and yellow) indicate the mapping of consumer chain heights to provider chain heights (note that on the provider chain there is only one block of a given color).  
```

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 5]
```diff
@@ -205,3 +213,62 @@ This approach is depicted in the following Figure that shows an overview of the
...
+- By default, every consumer CCV module maps any block height to `0` (i.e., VSC IDs start from `1`). 
+  Intuitively, this means that the voting power on the consumer chain at height `Hc` with `HtoVSC(Hc) = 0` was setup at genesis during Channel Initialization. 
+- For every consumer chain, the provider CCV module sets `VSCtoH[0]` to the height at which the first VSC was provided to this consumer chain. 
+  Intuitively, this means that the validator set on the provider chain at height `VSCtoH[0]` matches the validator set on the consumer chain at all heights `Hc` with `HtoVSC[Hc] = 0`.
+
+The following figure shows an overview of the Consumer Initiated Slashing operation of CCV. 
+
+![Consumer Initiated Slashing](./figures/ccv-evidence-overview.png?raw=true)
+
+- At height `Hc2`, the consumer chain receives evidence that a validator `V` misbehaved at height `Hc1`.
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:264]:
```suggestion
- At (evidence) height `Hc2`, the consumer chain receives evidence that a validator `V` misbehaved at (infraction) height `Hc1`. 
```

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 6]
```diff
@@ -205,3 +213,62 @@ This approach is depicted in the following Figure that shows an overview of the
...
+
+The following figure shows an overview of the Consumer Initiated Slashing operation of CCV. 
+
+![Consumer Initiated Slashing](./figures/ccv-evidence-overview.png?raw=true)
+
+- At height `Hc2`, the consumer chain receives evidence that a validator `V` misbehaved at height `Hc1`. 
+  As a result, the consumer CCV module sends a `SlashPacket` to the provider chain: 
+  It makes a request to slash `V`, but it replaces the infraction height `Hc1` with `HtoVSC[Hc1]`, 
+  i.e., the ID of the VSC that updated the "misbehaving voting power" or `0` if such a VSC does not exist.
+- The provider CCV module receives at height `Hp1` the `SlashPacket` with `vscId = HtoVSC[Hc1]`.
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:268]:
```suggestion
- The provider CCV module receives at (slashing) height `Hp1` the `SlashPacket` with `vscId = HtoVSC[Hc1]`. 
```

File: spec/app/ics-028-cross-chain-validation/system_model_and_properties.md [hunk 7]
```diff
@@ -169,6 +204,28 @@ The following properties define the guarantees of CCV on *registering* on the pr
...
+  Then, the CCV module of `cc` MUST send (to the provider CCV module) *exactly one* `SlashPacket` `P`, such that
+    - `P` is sent at height `h = max(he, hv)`;
+    - `P.val = val` and `P.id = HtoVSC[hi]`,
+    i.e., the ID of the latest VSC that updated the validator set on `cc` at height `hi` or `0` if such a VSC does not exist (if `hi < hv`).
+  > **Note**: A consequence of the *Consumer Slashing Warranty* property is that the initial validator set on a consumer chain cannot be slashed during the initialization of the CCV channel. 
+  Therefore, consumer chains *SHOULD NOT allow user transactions before the CCV channel is established*. 
+  Note that once the CCV channel is established (i.e., a VSC is received from the provider CCV module), CCV enables the slashing of the initial validator set for infractions committed during channel initialization.
+
+- ***Provider Slashing Warranty***: If the provider CCV module receives in a block from a consumer chain `cc` a `SlashPacket` containing a validator `val` and a VSC ID `vscId`,
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/system_model_and_properties.md:220]: I am not sure about the wording "in a block" in this property. Above "at height" is used which seems clearer to me.

**Review by josef-widder — COMMENTED** [review]
I specifically looked at the "Consumer initiated slashing" sections of the overview and the properties file. Apart minor comments, it looks great.

I will have a look at the corresponding section in the technical specification file later today.

File: spec/app/ics-028-cross-chain-validation/technical_specification.md [hunk 8]
```diff
@@ -1285,9 +1705,237 @@ function UnbondMaturePackets() {
...
+> **Note**: The ABCI application MUST subtract `ValidatorUpdateDelay` from the infraction height before invoking `SendSlashRequest`, 
+> where `ValidatorUpdateDelay` is a delay (in blocks) between when validator updates are returned to the consensus-engine and when they are applied 
+> (for more details, take a look at the [ABCI specification](https://github.com/tendermint/spec/blob/v0.7.1/spec/abci/abci.md#endblock)). 
+> For example, if `ValidatorUpdateDelay = x` and a validator set update is returned with new validators at the end of block `10`, 
+> then the new validators are expected to sign blocks beginning at block `11+x`. 
+> 
+> Consequently, the consumer CCV module expects the `infractionHeight` parameter of the `SendSlashRequest()` to be set accordingly.
+
+> **Note**: In the context of single-chain validation, slashing for downtime is an **_atomic operation_**, i.e., once the downtime is detected, the misbehaving validator is slash and jailed immediately. 
+> Consequently, once a validator is punished for downtime, is removed from the validator set and cannot be punished again for downtime.
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/technical_specification.md:1898]:
```suggestion
> Consequently, once a validator is punished for downtime, it is removed from the validator set and cannot be punished again for downtime. 
```

File: spec/app/ics-028-cross-chain-validation/technical_specification.md [hunk 9]
```diff
@@ -1285,9 +1705,237 @@ function UnbondMaturePackets() {
...
+  - The acknowledgement is `SlashPacketError`.
+
+<!-- omit in toc -->
+#### **[CCV-CCF-TOSLASH.1]**
+```typescript
+// CCF: Consumer Chain Function
+function onTimeoutSlashPacket(packet Packet) {
+  channelStatus = INVALID
+
+  // TODO What do we do here?
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/technical_specification.md:1813]: Is this related to https://github.com/cosmos/ibc/issues/669? Is this something we need to figure out immediately?

File: spec/app/ics-028-cross-chain-validation/README.md [hunk 10]
```diff
@@ -29,11 +29,14 @@ The communication between the provider and the consumer chains is done through t
 ## Backwards Compatibility
 
 (discussion of compatibility or lack thereof with previous standards)
+ -->
+
 
 ## Forwards Compatibility
 
-(discussion of compatibility or lack thereof with expected future standards)
+Future version may use newer versions of ABCI (e.g., ABCI++), which may result in different Tendermint block headers and, consequently, in different values for `ValidatorUpdateDelay` (see the [External Data Structures](./technical_specification.md#external-data-structures) section).
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/README.md]: I already removed this, i.e., https://github.com/cosmos/ibc/commit/8b0915beef3dd2778e30656d781ad6e10ec6b58c, since I decided that `ValidatorUpdateDelay` should be part of the Slashing module. There is a note about this in the Technical Spec. I'll add the Forwards Compatibility section once the entire draft is ready.

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 11]
```diff
@@ -205,3 +213,62 @@ This approach is depicted in the following Figure that shows an overview of the
...
+
+The following figure shows the intuition behind such a mapping using the provided VSCs. 
+The four unbonding operations (i.e., undelegations) occur on the provider chain and, as a consequence, the provider chain provides VSCs to the consumer chain, e.g., `undelegate-3` results in `VSC3` being provided.
+The four color (i.e., red, blue, green, and yellow) indicate the mapping of consumer chain heights to provider chain heights (note that on the provider chain there is only one block of a given color).  
+As a result, a validator misbehaving on the consumer chain, e.g., in either of the two green blocks, is slashed the same as if misbehaving on the provider chain, e.g., in the green block. 
+This ensures that once unbonding operations are initiated, the corresponding unbonding tokens are not slashed for infractions committed in the subsequent blocks, e.g., the tokens unbonding due to `undelegate-3` are not be slashed for infractions committed in or after the green blocks.
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:241]: The idea here is to explain via the diagram the intuition behind the height mapping. The period a validator is slashable is not needed for understanding the height mapping. Thus, I don't see why adding the unbonding period to this explanation would make it easier to understand.

File: spec/app/ics-028-cross-chain-validation/technical_specification.md (hunk 9 above)
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/technical_specification.md:1813]: Yes, it is related to https://github.com/cosmos/ibc/issues/669. There are opened issues for all the TODOs; I keep them also in the spec to emphasized that's something we are working on.

Comment (josef-widder) [reply spec/app/ics-028-cross-chain-validation/technical_specification.md:1813]: Great. In future project we might link the issue in the files for context. If that's too much overhead don't worry.

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 12]
```diff
@@ -205,3 +213,62 @@ This approach is depicted in the following Figure that shows an overview of the
...
+For the [Security Model](#security-model) to be preserved, misbehaving validators MUST be slashed (and MAY be jailed, i.e., removed from the validator set). 
+A prerequisite to slash validators is to receive valid evidence of their misbehavior. 
+Thus, when slashing a validator, we distinguish between three events and the heights when they occur:
+- `infractionHeight`, the height at which the misbehavior (or infraction) happened;
+- `evidenceHeight`, the height at which the evidence of misbehavior is received;
+- `slashingHeight`, the height at which the validator is slashed (and jailed). 
+
+> **Note**: In the context of single-chain validation, usually `evidenceHeight = slashingHeight`. 
+
+The [Security Model](#security-model) guarantees that any misbehaving validator can be slashed for at least the unbonding period,
```
Comment (jovankomatovic) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:229]: I would maybe rephrase this sentence as I understood it in the following way: "If any misbehaving validator misbehaves, it is slashed within the unbonding period".
Maybe: "The security model allows any misbehaving validator to be slashed only within the unbonding period".

Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:229]: That's the current behavior in Cosmos SDK. But this is not necessary for security. Validators can be slashed also after the unbonding period elapses and there will be no impact on security. In CCV, this may be the case since a validator must wait to unbond on all consumer chains. Thus, even though the unbonding period elapsed on the provider chain, the validator can still be slashed.

**Review by AdityaSripal — COMMENTED** [review]
Another partial review. Looking good

File: spec/app/ics-028-cross-chain-validation/system_model_and_properties.md [hunk 13]
```diff
@@ -64,47 +67,79 @@ The following assumptions define the guarantees CCV expects from the provider St
   Then,
   - (*unbonding initiation*) the provider CCV module MUST be notified of `uo`'s initiation before receiving `U(uo)`;
   - (*unbonding completion*) `uo` MUST NOT complete on the provider chain before the provider chain registers notifications of `vsc(uo)`'s maturity from all consumer chains.
+  > **Note**: Depending on the implementation, the (*unbonding initiation*) part of the *Unbonding Safety* MAY NOT be necessary for validator unbonding operations.
+
+- ***Slashing Warranty***: If the provider ABCI application (e.g., the Slashing module) receives a request to slash a validator `val` that misbehaved at block height `h`, then it slashes the amount of tokens `val` had bonded at height `h` except the amount that has already completely unbonded. 
 
-> **Note**: Depending on the implementation, the (*unbonding initiation*) part of the *Unbonding Safety* MAY NOT be necessary for validator unbonding operations.
+- ***Evidence Provision***: If the consumer ABCI application receives a valid evidence of misbehavior at block height `h`, then it MUST submit it to the consumer CCV module *exactly once* and at the same height `h`.
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/system_model_and_properties.md:74]: At the moment we don't track on the consumer which validators fully unbonded. For that we should keep a list of unbonding validators mapped to the VSC that set their power to 0. However, I don't think it's necessary, at least not for security. The consumer sends the SlashPacket, the provider receives it and request the staking module to slash. If the validator is already unbonded, nothing will happen.

File: spec/app/ics-028-cross-chain-validation/system_model_and_properties.md [hunk 14]
```diff
@@ -64,47 +67,79 @@ The following assumptions define the guarantees CCV expects from the provider St
...
-CCV provides the following system invariants:
+CCV provides the following system invariants. 
 - ***Validator Set Invariant***: Every validator set on any consumer chain MUST either be or have been a validator set on the provider chain.
-- ***Voting Power Invariant***: Let
-  - `pBonded(t,val)` be the number of tokens bonded by validator `val` on the provider chain at time `t`; 
-    note that `pBonded(t,val)` includes also unbonding tokens (i.e., tokens in the process of being unbonded);
-  - `VP(T)` be the voting power associated to a number `T` of tokens;
-  - `Power(cc,t,val)` be the voting power granted to a validator `val` on a consumer chain `cc` at time `t`.  
+
+- ***Voting Power Invariant***: Let `val` be a validator, `cc` be a consumer chain, both `hc` and `hc'` be heights on `cc`, and both `hp` and `hp'` be heights on the provider chain, such that
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/system_model_and_properties.md:106]: Not sure whether CCV requires this. That's a staking module requirement, right?

**Review by josef-widder — APPROVED** [review]
Looks great! Thanks for the work!!

**Review by AdityaSripal — APPROVED** [review]
Nice work!

File: spec/app/ics-028-cross-chain-validation/system_model_and_properties.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (AdityaSripal) [reply spec/app/ics-028-cross-chain-validation/system_model_and_properties.md:106]: Well yes, its encapsulated in the requirements of the staking module

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 15]
```diff
@@ -205,3 +213,62 @@ This approach is depicted in the following Figure that shows an overview of the
...
+
+The following figure shows an overview of the Consumer Initiated Slashing operation of CCV. 
+
+![Consumer Initiated Slashing](./figures/ccv-evidence-overview.png?raw=true)
+
+- At (evidence) height `Hc2`, the consumer chain receives evidence that a validator `V` misbehaved at (infraction) height `Hc1`. 
+  As a result, the consumer CCV module sends a `SlashPacket` to the provider chain: 
+  It makes a request to slash `V`, but it replaces the infraction height `Hc1` with `HtoVSC[Hc1]`, 
+  i.e., the ID of the VSC that updated the "misbehaving voting power" or `0` if such a VSC does not exist.
+- The provider CCV module receives at (slashing) height `Hp1` the `SlashPacket` with `vscId = HtoVSC[Hc1]`.
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:268]: Yes, I find that too much of a detail for the overview. That's described in the technical spec.

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 16]
```diff
@@ -205,3 +213,62 @@ This approach is depicted in the following Figure that shows an overview of the
   The unbonding operation is completed later once the provider CCV module registered maturity notifications for `VSC2` from all consumer chains.
 
 ![Completion of Unbonding Operations](./figures/ccv-unbonding-overview.png?raw=true)
+
+### Consumer Initiated Slashing
+[&uparrow; Back to Outline](#outline)
+
+For the [Security Model](#security-model) to be preserved, misbehaving validators MUST be slashed (and MAY be jailed, i.e., removed from the validator set).
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:220]: The actual requirement is "misbehaving validators MUST be punished". This punishment can come in the form of slashing and maybe jailing. IMO, slashing is mandatory. Jailing without slashing will not be enough (since the attacker may leave the set anyway after an attack). Slashing without jailing will still satisfy the safety requirement, since it will act as a deterrent agains misbehavior. I agree though that both are better.
